src/server/pps/server/api_server.go

```diff
@@ -502,6 +629,11 @@ func newPipelineInfo(persistPipelineInfo *persist.PipelineInfo) *ppsclient.Pipel
 func (a *apiServer) runPipeline(pipelineInfo *ppsclient.PipelineInfo) error {
 	ctx, cancel := context.WithCancel(context.Background())
 	a.cancelFuncsLock.Lock()
+	if _, ok := a.cancelFuncs[*pipelineInfo.Pipeline]; ok {
+		// The pipeline is already being run
+		a.cancelFuncsLock.Unlock()
+		return nil
+	}
```

Comment:
It seems like based on how `AddShard` and `DeleteShard` are implemented and the fact that calls to them can't overlap we'll never hit this. Is that true or am I missing something?